### PR TITLE
fix: update instrument, account for lack of arrow expression (#119)

### LINKED ISSUE #119: istanbul-lib-instrument: compatibility with babel@7
Since `@babel/core@7.0.0-beta.33`, this package started logging this statement in the [console](https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-instrument/src/visitor.js#L223):

`Unable to process function body node type: ArrowFunctionExpression`

According to [release notes](https://github.com/babel/babel/releases/tag/v7.0.0-beta.33), there have been some breaking changes, namely:

> removing expression field from ArrowFunctionExpression.

Used packages:

```
"@babel/core": "^7.0.0-beta.34"
"babel-plugin-istanbul": "^4.1.5" (which uses istanbul-lib-instrument)
```

Also let me know if I should move this issue over to [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul).

### PR DESCRIPTION
**Pull request by tswaters**
Fixes #119

### CODE REVIEW

**Comment by tswaters** [comment]
Really the two statements are equivalent; an `ArrowFunctionExpression` whose body type isn't BlockStatement is an expression.

Looking through the tests, I'm not sure to test this specific thing.  One would need to stub the return of babylon to include or not include the `expression` property.  

Any thoughts on how to effectively test this?

**Comment by lukescott** [comment]
Looks like the relevant test would be:

https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-instrument/test/specs/arrow-fn.yaml

Did the test fail before the fix? It looks like most of the tests are single-parameter (all?). The last case has `(bar) =>` vs `bar =>`. Does that cover both cases? If so, it would seem the cases are covered.  If not, would adding a 2 parameter test cover all cases?

I'm eager to get this fix merged as my test coverage has dropped since updating to Babel 7. I updated because I'm currently using/testing the TypeScript preset, in favor of tsc + Babel 6.

**Comment by tswaters** [comment]
> Did the test fail before the fix? 

No, but only because the babylon dependency is still the old version and all the tests run against it....  I could try to bump the dependency, but could have other unintended consequences (plus it's still beta).  The only change that is causing this particular problem is no longer including the `expression` property on `ArrowFunctionExpression` - the function can be fixed to handle both cases.

re: single parameter tests

That's a good call, I'll include more tests.  fwiw, the last test is only verifying the name of the function is set properly - the first few tests cover `x => x + x` and `x => return {x + x}`... still, I've added a few tests for multiple arguments just to be sure.

**Comment by lukescott** [comment]
You could update the dependency locally without the fix to see if the tests fail, then ensure they pass with the fix. That way when the dependency does get updated the tests will pass.

It would also be good to see if any other tests fail in other areas and get issues logged if there are any.

**Comment by tswaters** [comment]
I've verified locally -- `babylon@next` fails the tests.  Fortunately, it is just the arrow function, and under this branch the tests pass.

```
 1) arrow-fn.yaml/es6 arrow function expression function called:

      Line coverage mismatch
      + expected - actual

       {
         "1": 1
      -  "2": 1
      +  "2": 4
       }

      at Function.assert.deepEqual (node_modules\chai\lib\chai\interface\assert.js:216:32)
      at Verifier.verify (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/util/verifier.js:46:16)
      at Context.fn (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/specs.test.js:73:31)

  2) arrow-fn.yaml/es6 arrow function expression function not called:

      Statement coverage mismatch
      + expected - actual

       {
         "0": 1
         "1": 1
      +  "2": 0
       }

      at Function.assert.deepEqual (node_modules\chai\lib\chai\interface\assert.js:216:32)
      at Verifier.verify (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/util/verifier.js:49:16)
      at Context.fn (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/specs.test.js:73:31)

  3) arrow-fn.yaml/complex arrow fn call nested arrow fn:

      Line coverage mismatch
      + expected - actual

       {
         "1": 1
      +  "2": 1
      +  "3": 1
      +  "4": 1
         "8": 1
       }

      at Function.assert.deepEqual (node_modules\chai\lib\chai\interface\assert.js:216:32)
      at Verifier.verify (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/util/verifier.js:46:16)
      at Context.fn (D:/Code/istanbuljs/packages/istanbul-lib-instrument/test/specs.test.js:73:31)
```

**Comment by lukescott** [comment]
@JaKXz would it be possible to get this merged soon? I would pull from @tswaters branch for now, but it's difficult to pull in sub-packages from a monorepo.

File: packages/istanbul-lib-instrument/src/visitor.js [hunk 1]
```diff
@@ -343,7 +343,7 @@ function parenthesizedExpressionProp(prop) {
 function convertArrowExpression(path) {
     const n = path.node;
     const T = this.types;
-    if (n.expression) {
+    if (n.expression || n.body.type !== 'BlockStatement') {
```
[the remark this answers is not in the document]
Comment (tswaters) [reply packages/istanbul-lib-instrument/src/visitor.js]: I've updated this.

File: packages/istanbul-lib-instrument/src/visitor.js [hunk 2]
```diff
@@ -343,7 +343,7 @@ function parenthesizedExpressionProp(prop) {
...
         const bloc = n.body.loc;
         n.expression = false;
```
[the remark this answers is not in the document]
Comment (tswaters) [reply packages/istanbul-lib-instrument/src/visitor.js]: I've updated this.

**Comment by bcoe** [comment]
@tswaters awesome, thank for digging into this; I will try to get a release of this out early this week (along with @danez's changes); I just want to make sure I test thoroughly.

**Comment by VladShcherbin** [comment]
@bcoe hey, any chance to merge this one? It would be super cool to have arrow functions working.

**Comment by bcoe** [comment]
@VladShcherbin I've been dealing with a family tragedy for the past few months, I apologize for taking so long to merge this -- I'll work on having a release out soon for you to test with.

**Comment by bcoe** [comment]
@tswaters @VladShcherbin this is released now, please let me know if this does the trick for you.

**Comment by VladShcherbin** [comment]
@bcoe everything is working fine for me now, thank you 👍

**Pull request merged**